ec/src/models/short_weierstrass/affine.rs

```diff
@@ -0,0 +1,420 @@
+    // DISCUSS Maybe these function names are too verbose?
+    pub fn is_in_correct_subgroup_assuming_on_curve(&self) -> bool {
+        P::is_in_correct_subgroup_assuming_on_curve(self)
+    }
+}
+
+impl<P: SWCurveConfig> Zeroize for Affine<P> {
+    // The phantom data does not contain element-specific data
+    // and thus does not need to be zeroized.
+    fn zeroize(&mut self) {
+        self.x.zeroize();
+        self.y.zeroize();
+        self.infinity.zeroize();
+    }
+}
+
+impl<P: SWCurveConfig> Distribution<Affine<P>> for Standard {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Affine<P> {
+        loop {
+            let x = P::BaseField::rand(rng);
+            let greatest = rng.gen();
+
+            if let Some(p) = Affine::get_point_from_x(x, greatest) {
+                return p.mul_by_cofactor();
+            }
+        }
+    }
+}
+
+impl<P: SWCurveConfig> AffineRepr for Affine<P> {
+    type Config = P;
+    type BaseField = P::BaseField;
+    type ScalarField = P::ScalarField;
+    type Group = Projective<P>;
+
+    fn xy(&self) -> Option<(&Self::BaseField, &Self::BaseField)> {
+        (!self.infinity).then(|| (&self.x, &self.y))
+    }
+
+    #[inline]
+    fn generator() -> Self {
+        P::GENERATOR
+    }
+
+    fn identity() -> Self {
+        Self {
+            x: P::BaseField::ZERO,
+            y: P::BaseField::ZERO,
+            infinity: true,
+        }
+    }
+
+    fn from_random_bytes(bytes: &[u8]) -> Option<Self> {
+        P::BaseField::from_random_bytes_with_flags::<SWFlags>(bytes).and_then(|(x, flags)| {
+            // if x is valid and is zero and only the infinity flag is set, then parse this
+            // point as infinity. For all other choices, get the original point.
+            if x.is_zero() && flags.is_infinity() {
+                Some(Self::identity())
+            } else if let Some(y_is_positive) = flags.is_positive() {
+                Self::get_point_from_x(x, y_is_positive)
+                // Unwrap is safe because it's not zero.
+            } else {
+                None
+            }
+        })
+    }
+
+    fn mul_bigint(&self, by: impl AsRef<[u64]>) -> Self::Group {
+        P::mul_affine(self, by.as_ref())
+    }
+
+    /// Multiplies this element by the cofactor and output the
+    /// resulting projective element.
+    #[must_use]
+    fn mul_by_cofactor_to_group(&self) -> Self::Group {
+        P::mul_affine(self, Self::Config::COFACTOR)
+    }
+
+    /// Performs cofactor clearing.
+    /// The default method is simply to multiply by the cofactor.
+    /// Some curves can implement a more efficient algorithm.
+    fn clear_cofactor(&self) -> Self {
+        P::clear_cofactor(self)
+    }
+}
+
+impl<P: SWCurveConfig> Neg for Affine<P> {
+    type Output = Self;
+
+    /// If `self.is_zero()`, returns `self` (`== Self::zero()`).
+    /// Else, returns `(x, -y)`, where `self = (x, y)`.
+    #[inline]
+    fn neg(mut self) -> Self {
```

Comment:
Just a confirmation that we are on the same page that if one does `negq = q.neg()`, an implicit copy happens, and q will remain unchanged, right? @confuseSUN mentioned this to me a few hours ago.

Reply:
Yes, it’s not a `neg_in_place` (and in fact there’s no such operation)